### PR TITLE
revise shell contact constructor

### PR DESCRIPTION
**Pull request by Xiangyu-Hu**
Hi, Dong,
please check my revision on the constructor and compared the the differences.
Xiangyu

### CODE REVIEW

**Comment by DongWuTUM** [comment]
Prof. Hu, 

The simplified constructor is well orginazed and more efficient. I should learn form you.

I found a bug (caused by me) when calculating `contact_max_`. I have corrected it and add regression test for case of 3d_elasticSolid_shell_collision. After corrected bug, the `heuristic_limiter` can be set as 1.0, i.e, this limiter is not needed anymore. But I still keet it and set it as 0.4 as it do improve the stability.

Dong

**Comment by DongWuTUM** [comment]
@BenceVirtonomy @FabienPean-Virtonomy 
Hi, Bence and Fabien, 
I found a bug when calculating contact_max_. After that, the heuristic_limiter can be set as 1.0, i.e, this limiter is not needed anymore. But I still keet it and set it as 0.4 as it do improve the stability. Sorry for misleading.
Dong

**Comment by BenceVirtonomy** [comment]
>

**Comment by BenceVirtonomy** [comment]
> @BenceVirtonomy @FabienPean-Virtonomy Hi, Bence and Fabien, I found a bug when calculating contact_max_. After that, the heuristic_limiter can be set as 1.0, i.e, this limiter is not needed anymore. But I still keet it and set it as 0.4 as it do improve the stability. Sorry for misleading. Dong

Does that mean that it should be calibrated on a test case basis? Meaning different scenarios might require a different constant?

File: SPHINXsys/src/shared/particle_dynamics/solid_dynamics/contact_dynamics.cpp [hunk 1]
```diff
@@ -77,52 +77,35 @@ namespace SPH
...
-							(kernel_->W(contact_h_ratio_[k], three_gaussian_points_[i] * contact_particle_spacing_[k] * 0.5 + contact_particle_spacing_[k] * 0.5, Vec3d{0,0,0})
-								- offset_W_ij_[k])
-							* 2.0 * Pi * (three_gaussian_points_[i] * contact_particle_spacing_[k] * 0.5 + contact_particle_spacing_[k] * 0.5)
-							* contact_particle_spacing_[k] * 0.5 * three_gaussian_weights_[i];
-					}
+					Real temp = three_gaussian_points_[l] * average_spacing_k * 0.5 + average_spacing_k * 0.5;
+					contact_temp = 2.0 * (kernel_->W(h_ratio_k, temp, ZeroVecd) - offset_W_ij_[l]) *
+						average_spacing_k * 0.5 * three_gaussian_weights_[l];
+					contact_max += contact_temp;
+					if (Dimensions == 3)
```
Comment (Xiangyu-Hu) [inline SPHINXsys/src/shared/particle_dynamics/solid_dynamics/contact_dynamics.cpp:102]: should it be contact_max += Dimensions == 3 ? contact_temp : 	contact_temp * Pi * temp ?

Comment (DongWuTUM) [reply SPHINXsys/src/shared/particle_dynamics/solid_dynamics/contact_dynamics.cpp:102]: > should it be contact_max += Dimensions == 3 ? contact_temp : contact_temp * Pi * temp ?

Yes, better. Thank you. Seems like the regression test for 3d_elasticSolid_shell_collision has the problem. I will correct that together.

**Comment by DongWuTUM** [comment]
> > @BenceVirtonomy @FabienPean-Virtonomy Hi, Bence and Fabien, I found a bug when calculating contact_max_. After that, the heuristic_limiter can be set as 1.0, i.e, this limiter is not needed anymore. But I still keet it and set it as 0.4 as it do improve the stability. Sorry for misleading. Dong
> 
> Does that mean that it should be calibrated on a test case basis? Meaning different scenarios might require a different constant?

No, it should remain constant for all cases.

**Comment by DongWuTUM** [comment]
I will try to generate regression test data in Linux system.

**Comment by Xiangyu-Hu** [comment]
> I will try to generate regression test data in Linux system.

Have you run the test on linux system?

**Comment by DongWuTUM** [comment]
> > I will try to generate regression test data in Linux system.
> 
> Have you run the test on linux system?

Yes, I have.

**Comment by Xiangyu-Hu** [comment]
> > > I will try to generate regression test data in Linux system.
> > 
> > 
> > Have you run the test on linux system?
> 
> Yes, I have.

Why did you cancel the CI in linux, as it looks ok?

**Comment by DongWuTUM** [comment]
> > > > I will try to generate regression test data in Linux system.
> > > 
> > > 
> > > Have you run the test on linux system?
> > 
> > 
> > Yes, I have.
> 
> Why did you cancel the CI in linux, as it looks ok?

I think the Linux CI is cancelled by Github automatically, as it has run for more than 360 min.

**Comment by Xiangyu-Hu** [comment]
> > > > > I will try to generate regression test data in Linux system.
> > > > 
> > > > 
> > > > Have you run the test on linux system?
> > > 
> > > 
> > > Yes, I have.
> > 
> > 
> > Why did you cancel the CI in linux, as it looks ok?
> 
> I think the Linux CI is cancelled by Github automatically, as it has run for more than 360 min.



> > > > > I will try to generate regression test data in Linux system.
> > > > 
> > > > 
> > > > Have you run the test on linux system?
> > > 
> > > 
> > > Yes, I have.
> > 
> > 
> > Why did you cancel the CI in linux, as it looks ok?
> 
> I think the Linux CI is cancelled by Github automatically, as it has run for more than 360 min.

Which suggests that the two-phase dambreak case runs into a dead loop?

**Comment by DongWuTUM** [comment]
It looks like a dead loop.

**Comment by Xiangyu-Hu** [comment]
> It looks like a dead loop.
It passed now. So you can approve the pull request.

**Review by DongWuTUM — APPROVED** [review]

**Pull request merged**